src/native/minipal/cpufeatures.c

```diff
@@ -475,6 +503,12 @@ int minipal_getcpufeatures(void)
     // FP and SIMD support are enabled by default
     result |= ARM64IntrinsicConstants_AdvSimd;
 
+    // RDM does not have an IsProcessorFeaturePresent flag, but it is a requirement for Windows 11
+    if (IsWindows11OrGreater())
```

Comment:
We use https://github.com/search?q=repo%3Adotnet%2Fruntime%20IsWindowsVersionOrGreate&type=code for this purpose in other places

Reply:
`IsWindowsVersionOrGreater` also lies about Windows version and says 8.1 without a manifest. Using it will return invalid values with custom hosts.

Reply:
> IsWindowsVersionOrGreater also lies about Windows version and says 6.2 without a manifest. Using it will return invalid values with custom hosts.

This is intentional on the part of Windows and is covered under the `Remarks` section of the deprecated `GetVersion`/`GetVersionEx` functions and the recommended `VersionHelper` replacements: https://learn.microsoft.com/en-us/windows/win32/api/versionhelpers/

Applications must be explicitly manifested for Windows 10+, this manifest is likewise required for functions to use newer behaviors, for some (but not all) newer exports to be available, etc.

This is also required to ensure that user-selected comapt settings are respected, among other considerations. Trying to manually determine the actual underlying OS version is never recommended and liable to cause undefined or other problematic behaviors.

Reply:
@jkotas, notably the managed `OperationgSystem.IsWindowsVersionAtLeast` check looks to use `Environment.OSVersion.Version` which is currently populated via `RtlGetVersion`

This does not correctly respect the manifest file and will mismatch from the underlying `GetVersionEx` call and the corresponding `IsWindowsVersionOrGreater` calls that are currently recommended for use.

Is that known/intentional? It seems  like it should be unintentional and like it could cause various bugs/issues, especially where application behaviors may not match up...

Reply:
> Is that known/intentional? It seems like it should be unintentional and like it could cause various bugs/issues, especially where application behaviors may not match up...

https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/5.0/environment-osversion-returns-correct-version

Reply:
Yes, I know that `Environment.OSVersion` is intended to do that

However, `OperatingSystem.IsWindowsVersionAtLeast` using that seems incorrect and can potentially lead to problems due to mismatched behaviors, assumptions of APIs existing that may not be accessible, etc.

Reply:
Developer expectations mean little when the underlying OS requires something for correctness and documents that you must use the correct API for the right behaviors to exist.

Reply:
> managed OperationgSystem.IsWindowsVersionAtLeast check looks to use Environment.OSVersion.Version which is currently populated via RtlGetVersion
> Is that known/intentional?

Early in .NET Core days, we were respecting the Windows manual and returned the shimmed version from Environment.OSVersion. It resulted in endless stream of issues since people do not ever expect the API to return the shimmed version. We gave up and fixed the API to return the actual version.

I agree with you that the current behavior of IsWindowsVersionAtLeast violates the Windows manual. I think that fixing it to follow Windows manual would result in a confusion and unhappiness again. The current behavior does not seem to cause any problems in practice. Windows does not seem to use the manifested version as a tool to evolve the OS these days.

Reply:
> We use https://github.com/search?q=repo%3Adotnet%2Fruntime%20IsWindowsVersionOrGreate&type=code for this purpose in other places

Hmmm, I am wondering whether these checks are broken in practice since we do not have Win 10 in the win32 manifest enabled by default (https://github.com/dotnet/sdk/issues/12904).

Reply:
> Windows does not seem to use the manifested version as a tool to evolve the OS these days.

I'm not so sure; the documentation is still semi-regularly updated and they've explicitly stated that from Windows 8+ (when they made these manifest changes) they started taking a proactive, rather than reactive, approach to compatibility. Some of this is covered in the general compatibility cookbook: https://learn.microsoft.com/en-us/windows/compatibility/ -- Noting that Windows 11 (despite the name) is still classified as v10.0 at the kernel level and so is often grouped together in various manifest documentation.

And while for most new APIs you can resolve them without a manifest, this is really on a case-by-case basis. There are some where this isn't true and a manifest of some form is required (a slightly older, but well known example is `comctl32` and visual styles: https://learn.microsoft.com/en-us/windows/win32/controls/cookbook-overview)

It might be a good idea to sync with the Windows team and get some more information on the official guidance, plans, and/or impact of the manifest for the future and the potential impact not respecting it may have. -- At the very least it seems very odd to do one thing for the managed side and a different thing for the native side. In this case without a manifest the VM will say RDM is unsupported but the managed side will tell the dev they are on Win 11 and therefore it should be.

Reply:
> Hmmm, I am wondering whether these checks are broken in practice since we do not have Win 10 in the win32 manifest enabled by default (https://github.com/dotnet/sdk/issues/12904).

I believe they are broken.

I didn't test the IsWindowsVersion or greater locally because I assumed it worked based on the fact it was relied on elsewhere, but it doesn't work.  The RtlGetVersion code did.

Reply:
I'll also point out that the original idea of imlpying RDM from DP had an implicit Windows 11 dependency because the DP flag only exists in Win11.  That could leave out some supported hardware in theory, but I don't think it affects any actual Windows-compatible hardware.

> Support for PF_ERMS_AVAILABLE, PF_ARM_V82_DP_INSTRUCTIONS_AVAILABLE, and PF_ARM_V83_JSCVT_INSTRUCTIONS_AVAILABLE were added in the Windows SDK (20348) and are supported by Windows 11 and Windows Server 2022.

Reply:
> DP had an implicit Windows 11 dependency because the DP flag only exists in Win11

This assumes that the support was not backported to Windows 10. My personal preferences starting from the most preferred:

I do not have a preference between `RtlGetVersion` and implying RDM from DP (with a good comment). I think it would be even acceptable to set the flag unconditionally `result |= ARM64IntrinsicConstants_AdvSimd | ARM64IntrinsicConstants_Rdm;` with the premise that it is part of Win11 baseline and that Win10 is out of support for .NET 10.

Reply:
> I believe they are broken.

Notably I don't think broken is the right word here. They work as Windows intends the APIs to work, which means the correct manifest must be set. .NET has a tracking issue around include a manifest by default today, which means it won't light up without the user doing something explicit; but that's a different issue (https://github.com/dotnet/sdk/issues/12904).

I think that `IsWindowsVersionOrGreater` is still the most correct option based on currently available docs. I would expect if we talked with the Windows team they'd indicate the same, but if that is no longer the case then it might influence the decision. 

`RtlGetVersion` seems is like the next safest/most desirable option and would allow it to lightup even in compatibility mode (assuming that the OS isn't doing something to disable such an ISA in compat mode, which is probably unlikely).

Hardcoding it under the presumption we are Windows 11+ seems like the least desirable option. Although its "technically ok", there will still be LTSC versions of Win10 supported for a few more years and likely users who refuse to move off of Windows 10 and who continue using it anyways. Not officially supporting it, but also not explicitly blocking it (like we did with Win7) seems like "goodness" then. -- If we did take such a dependency on Windows 11, however; then I'd really want to see us update the `BaselineIsa` targeted by NAOT, Crossgen, and other scenarios so we don't have as many opportunistic checks and can generate better code by default. We could likewise remove some of the `FEAT_LSE` (atomics) checks we have in the GC today under that presumption and so on.

Reply:
I have a strong preference for following the implication from Arm's documented requirements here.

In the end, I really just want this to work, so I'll do whatever gets this landed, but it seems to me that the alternatives are more fragile.

`IsWindowsVersionOrGreater` is broken in too many places to be considered a good fix.  It requires external changes in order to get Windows 11 to actually respond correctly to the question "are you at least Windows 11?".  Even if the SDK were fixed to manifest the default app host, `corerun` would still be broken, and there's an unknown tail of other things.

`RtlGetVersion` is less fragile but apparently still can be impacted by compatibility modes.

The number of never-gonna-happen things that would have to happen in order to make the DP->RDM implication not safe or reliable is just unimaginably small.

Reply:
-- Getting the SDK manifest fixed by default should really also be a trivial fix and is something we should be doing anyways. It's likely something we should've taken the few days to do 3 years ago and should have been the resolution to `Environment.OSVersion` rather than changing it to use `RtlGetVersion`.

A manifest being set for a .NET Project just requires `<ApplicationManifest>app.manifest</ApplicationManifest>`. We can trivially add to the SDK something like `<ApplicationManifest Condition="'$(ApplicationManifest)' == '' and $([System.OperatingSystem]::IsWindows())>app.manifest</ApplicationManifest>` which mirrors the manifest currently used by `dotnet.exe` and `corerun.exe` (`longPaths=true`, `supportedOS=win7, win8, win10`).

We could optionally make this more robust; integrating friendly msbuild options for things like LongPaths, HighDpi, UAC settings, osplatform, etc; but that is additional/extra and not necessary; especially to resolve the bigger problem.

Without such a manifest; some behaviors like LongPaths, HighDpi support, and other general OS behaviors not match the documented Windows 11 behaviors (they may be quirked to Vista or Win 8 compatibility; currently Task Manager indicates Vista compatibility). Without the manifest some APIs may not be resolvable despite OSVersion  and the managed IsWindowsVersionAtLeast reporting Win11 is the target, and so on.

Once we have the manifest; then the C `IsWindowsVersionOrGreater` API is trivially correct. The `Environment.OSVersion` API would also be trivially correct for the `GetVersionEx` API that exists for client apps (unlike `RtlGetVersion` which is an "internal" API primarily meant for driver development and which isn't available to the client Windows SDK by default).

Everything mostly falls out if we fix the overarching and underlying issue of the manifest not being consistent and the entire ecosystem likely gets a little bit better and more stable. Developers can still override this manifest with their own or explicitly target a different OS Context via the exe file properties/registry entry.

Reply:
> IsWindowsVersionOrGreater is broken in too many places to be considered a good fix. It requires external changes in order to get Windows 11 to actually respond correctly to the question "are you at least Windows 11?

This isn't broken; it is an intentional design point of the API because the OS Context the app runs under is not Windows 11 if the manifest is not set. You can and will get differing behaviors.

Some examples of these quirks are:
* you get a private thread pool, rather than the NT thread pool for RPCs.
* you get a potential race condition for `GetOverlappedResult` calls (which we use for System.IO, named pipes, and other scenarios)
* you get a different behavior for theming under high contrast mode (can impact accessibility)
* you may see different HighDPI behaviors due to newer per thread and per monitor awareness options
* you may see different behavior for some APIs that support long paths
* etc

Reply:
> Getting the SDK manifest fixed by default should really also be a trivial fix and is something we should be doing anyways

I don't disagree that this should be considered, but 1) it sounds like it's a potential breaking change, and 2) it will never be as reliable a fix for this specific issue as the other two options.

> This isn't broken

An API that doesn't correctly answer the question in its name is fundamentally broken.  I understand how manifests and compatibility modes work, but those concerns have nothing to do with "does this hardware support these instructions?", and adding them into the mix is unnecessary.

The current state of things is incredibly confusing from a dev user perspective.  `ilc --instruction-set native" will compile my code without RDM support if I run it on Windows and with it if I run in WSL.

Even more confusing is the fact that I can either add `--instruction-set rdma` to compile with RDM support or just use the intrinsics without `IsSupported` guards and then run that AOTed code just fine on Windows, while JITted code in the same environment throws PNSE.

Trading one set of confusing behaviors for another that adds manifests into the mix is not a good direction.

Ultimately this comes down to whether the alternative implementation introduces an actual compat risk, and I genuinely believe based on everything said so far that it does not.

Reply:
> but 1) it sounds like it's a potential breaking change

It's largely a correctness fix, especially since .NET already has the right manifest for some hosts. There are some potential breaks, but no more than other other behavioral fix we get across releases. Both developers and users then have multiple override options if targeting the latest OS context isn't the right answer for their app.

> and 2) it will never be as reliable a fix for this specific issue as the other two options.

It is 100% reliable and works exactly as Windows intends it to work. It works exactly like the docs officially recommend doing version checks. It works exactly like you would get in C/C++, which is that it matches behavior to the underlying manifest file.

> An API that doesn't correctly answer the question in its name is fundamentally broken.

You have misunderstood the API and what the question in its name means.

The API is not designed to answer the question of "what is the underlying operating system version the application is literally running on?"; it is instead designed to answer the question of "what is the operating system version the application is running on?"

These sound similar but are quite different in the context of the OS and its design. The OS itself ships with multiple levels of compatibility, shimming, and other support. With this design, APIs are answering questions about the "OS Context" that the application is running in and therefore what OS the application should consider itself to be targeting, regardless of what the actual underlying system is. This is relevant for many reasons, such as to allow for bug fixes and behavior changes to an OS that runs on billions of devices worldwide and where most apps continue "just working" despite major architectural changes, rewrites, fixes, etc that have occurred over the past 32+ years.

Not everything works perfectly of course and some details do leak through, but for the most part they do. This is no different than a 32-bit application querying `C:\Program Files\...` and having it automatically redirect to `C:\Program Files (x86)\...` or a similar thing happening to registry queries or to `System32->SysWoW64`. It is no different than the general support that is required in order for x64 applications to run on an Arm64 device; or that ensures that a machine which is checking that if the name starts with `Windows 9` that it must be `Windows 95` or `Windows 98` and so on.

The applications and libraries that tend to break in these scenarios and not work compatibly are the ones doing exactly what you're suggesting; to bypass the documented and recommended check and to try and force query the underlying OS for additional information. You can of course do that and there are some scenarios where that is appropriate, but it introduces additional risk and potential for failure.

> The current state of things is incredibly confusing from a dev user perspective. ilc --instruction-set native will compile my code without RDM support if I run it on Windows and with it if I run in WSL.

Many of these options are technically undocumented and not fully supported today. They exist for power users who more or less know what they're doing.

Many instruction sets require support of the underlying OS, including some like AVX (which doesn't work pre-Windows 7), AVX512, APX, or even SVE on Arm64, etc. Some of these (SVE in particular) can even have additional options that can be controlled per-process and where ignoring the OS context is then problematic and can lead to downstream failures, because there is a difference between the hardware support, the OS support, and the application support.

> Even more confusing is the fact that I can either add --instruction-set rdma to compile with RDM support or just use the intrinsics without IsSupported guards and then run that AOTed code just fine on Windows, while JITted code in the same environment throws PNSE.

This can equally end up in a scenario where it fails. You can target `--instruction-set sve` and have it fail on one build of Windows 11 and succeed on a different build on the same hardware. 

The JIT targets the machine it detects and it detects that according to what the OS documents and provides to us. AOT targets the machine you tell it to and may cause hardware level faults if it is used incorrectly. -- It notably should be validating the used ISAs against the detected ISAs and fail to launch if they mismatch, if it isn't that sounds like a bug and something we may need to fix or investigate.

> Trading one set of confusing behaviors for another that adds manifests into the mix is not a good direction.

I strongly disagree with the premise that it is a different set of confusing behaviors as it is the documented way to do such OS version queries. Additionally, such an OS version query is the official recommendation that we got from the Windows team when a thread was opened requesting that an explicit FEAT_RDM flag be added.

This is the same behavior that a developer targeting C/C++ would be expected to follow to determine support. It is the same behavior that a developer targeting Rust or Java would be expected to follow.

It is the same behavior that you are expected to follow to ensure that `GetOverlappedResult` behaves correctly and doesn't have a race condition. It is the same behavior you are expected to follow in order for `LongPaths` to reliably work (which notably also means that despite .NET opting into and presuming LongPaths work, because its missing from the default manifest for the host with published apps; LongPaths may not actually work as expected).

It is the behavior everyone is supposed to use and follow and so it remains the most correct and least problematic behavior for .NET to use. It is the most likely to lead to success both short and long term.

Reply:
Your description of the intended design of the Windows OS version shimming feature based on app manifests is accurate. In practice, this feature did not seem to be successful and appears to be abandoned. In its golden days (2007 - 2015), a new OS version GUID was introduced about every 2 years. No new OS version GUID was introduced in last 9 years. For OS, it is hard to maintain and test the side-by-side behaviors like that at scale. We have learned our lesson by trying to adopt this approach in .NET Framework. For users, it is hard to reason about the matrix of manifest versions and actual versions. Also, the manifests work poorly with plugins.

> It is the same behavior that a developer targeting Rust or Java would be expected to follow.

It is a fine theory, but it is not what happens in practice:

Rust OS info crate uses [RtlGetVersion](https://github.com/stanislav-tkach/os_info/blob/d824f1ff58f120cd11256579d0acb84d1f7e3c30/os_info/src/windows/winapi.rs#L137)
Java is even more hard core and [gets the Windows OS version by reading the version of kernel32.dll](https://github.com/openjdk/jdk/blob/484229e04b812acd0c58a261c935c9e9190b3ba8/src/hotspot/os/windows/os_windows.cpp#L4002). I assume that it is to avoid the maze of version lies from the various APIs.

We have done the same in number of places. The most confusing part is .NET Framework where GetVersionEx and similar APIs are typically shimmed by Windows to return the actual OS version without respecting manifest versions for better compatibility.

Reply:
> In practice, this feature did not seem to be successful and appears to be abandoned

@jkotas I'm not sure it's abandoned because it appears there are still new individual manifest options being added in Win10/11 builds. What appears to be the case is rather that they've opted to not add a new `compat layer` to Windows since Win10. This looks to follow their existing practice which is that they've only added new layers for Major/Minor version updates. Since Win11 is still v10.0, it tracks that no new layer was added for it. My guess is then that if they ever update the kernel to `v10.1` or to `v11.0` then we will get a new `supportedOS Id="{guid}"` entry at that time.

I also believe that regardless of whether it is effectively abandoned moving forward or not; it appears as though we still require a manifest for all quirks and/or options introduced between Vista through Windows 10 to work correctly. This includes things like [LongPath](https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry#enable-long-paths-in-windows-10-version-1607-and-later), native control theming and [visual styles](https://learn.microsoft.com/en-us/windows/win32/controls/cookbook-overview), [HighDPI Awareness](https://learn.microsoft.com/en-us/windows/win32/hidpi/setting-the-default-dpi-awareness-for-a-process) (which while you can set it programmatically, this isn't recommended and fails in many scenarios if any HWND is created first), there are a number of additional documented compat layer quirks [here](https://learn.microsoft.com/en-us/windows/compatibility/application-executable-manifest) which includes a known race condition for `GetOverlappedResult` if you haven't opted into the Win7+ layer (Vista is the current default, as per the OS Context in TaskMgr and other locations).

Once that manifest file exists; then the question of using RtlGetVersion vs IsWindowsVersionOrGreater essentially goes away as we're following the best practices, opting into the relevant fixes for the latest version of .NET, and the version APIs do the right thing by default.

----------

I think the right next step here is to open a new internal thread with the Windows team (if this is the same person as the last thread, I'd be happy to start it or continue since its still related to the past thread). This will allow us to get an official answer both with regards to whether manifests are still needed/recommended for apps today to get the correct behavior and if there are any plans to continue versioning them in the future.

Getting confirmation will avoid any speculation and give us a clear direction to ensure we're not backing ourselves into a corner or causing potential issues here. It will also help inform us the importance of getting a "correct" manifest by default into the .NET SDK and may help direct us on how to handle user bug reports that exist around such scenarios.

Reply:
I agree that we should fix the manifests for Windows. I was trying to point out that the Windows developers at large try to avoid the OS version shimming since it does not work well for them.

> Once that manifest file exists; then the question of using RtlGetVersion vs IsWindowsVersionOrGreater essentially goes away 

It does not go away completely since manifests do not work well for plugins.

Reply:
> It does not go away completely since manifests do not work well for plugins.

Not sure I understand this?

Plugins and libraries in general are at the mercy of the application as it decides the OS context under which code runs. If the app doesn't include a manifest then IsWindowsVersionOrGreater correctly reports that the OS Context under which it is running and things like RDM or other functionality detailed above may not be available. As far as the app/library/plugin is concerned, unless it intentionally digs deeper, it's running on 6.2 -- Windows 8.

If the app does include a manifest, then the same APIs reports the lower of the greatest listed OS that is supported and the actual underlying OS being run on. This is the general target that the code should consider that it is running against and so everything continues to work.

A plugin or library could "require" that they run on a particular version or later and there isn't a way for them to force an application to use such a context; but they can add a relevant DllMain (C/C++) or Module Initializer/Static Constructor (.NET) which validates this and fails without -- most often, however, its better/easier to just not light up on the optional functionality. -- .NET assemblies also have the option to use `SupportedOSPlatform` and/or `TargetFramework` (`net8.0-windows10.0.22000.0` as an example) to further help enforce such usage.

Reply:
The problem is that the OS compat shims only isolate you from OS-version specific compat problems that were decided to be important enough to be shimmed. Many real-world OS-version specific compat issues are not shimmed and you are on your own. If you have to know the actual OS version to implement OS version specific workaround, you have to be able to get around the version lie shims.

The case study close to home is .NET Framework. Some OS version checks in .NET Framework had to get the real OS version and avoid version lie shims. Things would be broken otherwise. You can see some of that code in the initial .NET Core payload that included .NET Framework specifics: https://github.com/dotnet/runtime/blob/2024b63acd460edeb53b2ce6984710ab51c11f2f/src/coreclr/src/vm/ceemain.cpp#L4940-L4944

Reply:
I've reverted this back to the original implication and added a big comment.

Given the oldest processor supported on Windows 11 is based on the Arm Cortex-A76 core (and its small core sibling Cortex-A55), which implements both RDM and DP, I don't believe there is any chance of either false positive or false negative in any supported scenario.

I also believe this implication is, for all intents and purposes, identical to the implication that "we are running on Windows 11, therefore we have hardware that meets the minimum requirements for Windows 11" but without the surprising difficulty of answering the question "are we really running on Windows 11?"

Reply:
I did one last round of reading and wanted to add one final context note:

There is a single processor model on the [Windows 11 support list](https://learn.microsoft.com/en-us/windows-hardware/design/minimum/supported/windows-11-supported-qualcomm-processors) that might implement RDM but not DP: the Snapdragon 850.  That model has Cortex-A75 large cores and Cortex-A55 small cores.  I don't have one to test, but docs seem to indicate that the A75 does not implement DP, while the A55 does.  More commonly, the A55 appears as the small core alongside the Cortex-A76, which also supports both RDM and DP.  The A76/A55 combination is what's in the Surface X 2019 (Microsoft SQ1) model, Snapdragon 7c, and the Broadcom SoC used in Raspberry Pi 5.

Windows 11 does not actually require Armv8.1 until 24H2, according to https://www.neowin.net/news/first-generation-windows-on-arm-pcs-will-not-be-able-to-run-windows-11-24h2/.  There are reports that people have been running Windows 11 on unsupported devices with older builds and that 24H2 actually fails to boot.

So the DP->RDM implication includes a possible false negative on one supported processor model, while the Windows11->RDM implication includes possible false positives in unsupported configurations on older builds.

False negatives being preferable to false positives, I believe the implementation here is still right, but I can update the comment if it's worth churning the CI to get the exception noted there.